### ISSUE TITLE
Something weird in padding+shuffling

### ISSUE THREAD

**Issue by DenisPeskov** [pytorch/text#303]
Hello,

While building RNN models, I ran into a totally unexpected issue. 
Given working LSTM (or GRU) code, I had correctly decreasing train loss but increasing val loss.

ALL THINGS BEING EQUAL, changing batch_size from 32 to 4 caused the val loss to move in the correct direction and actually train.  I'm re-running with batch_size = 1 to confirm, but am pretty certain.

This seems potentially related to this "closed" issue.
https://github.com/pytorch/text/issues/70

Is there anything that would cause val/dev to not be properly sorted in larger batches?  My hypothesis is that sentences would then be extremely padded.  e.g. if you put a sentence of length 100 and 5 together, the latter would be padded 95 times.  Another reason might exist due to weird shuffling.  

If this hypothesis is correct, is there a way to fix this (ensure val/dev are sorted the same logical way train is)?  Hopefully a simple flag in BucketIterator or something similar.  
Will answer follow-up ?s or provide code if needed.

**Comment by DenisPeskov** [comment]
The last batches in particular have massive variations across train/val/dev.  

Most batches are within 1 in length, but one ranges in lengths from 64 to 180.  Another ranges from 61 to 155.

**Comment by DenisPeskov** [comment]
Issue 

packed = pack_padded_sequence(embed, list(lengths_cpu),
                                      batch_first=True)
output, self.hidden = self.rnn(packed)
x = self.hidden[0]

**Comment by mttk** [comment]
From what I can see, the issue is probably due to the [pool method](https://github.com/pytorch/text/blob/master/torchtext/data/iterator.py#L271) which uses a fixed size length lookahead to sort the bathces. 

Essentially, moving `lookahead * batch_size` examples from (train/test/val) set, sorts them and generates batches of approximately same length. So if you perhaps have a dataset of size (lookahead + 1), the last batch is going to be taken as-is (even if the examples are of length 5, 100 and 1000).

Apart from doing what you said (using packed sequences), you can probably set the `padding_idx` in the `nn.Embedding` to 1 (which is the torchtext default) and *you should* get similar results. 

Alternatively (and probably best), set batch_size to `1` in test / valid.

**Comment by DenisPeskov** [comment]
Thanks, nonissue now since I figured it out.  

I had padding_Idx set (so those embeddings were all 0s) and it produced junk.  

batch_size seems like an easy change.  Both would take notably longer on large datasets with batch of 1 though (but unless data is in the order of millions, a matter of extra minutes not hours).

**Comment by mttk** [comment]
Yeah -- just wanted to give you another option. It will make things slower, but in datasets of size of orders of million, you'd expect this problem not to influence the total loss by a lot. Sorry for being slow to respond, got a lot on my plate atm.
